# 1. Carregamento do dataset

O dataset Iris foi carregado utilizando a biblioteca `scikit-learn` e
convertido para um DataFrame do Pandas, facilitando sua análise e
pré-processamento.

In [24]:
import pandas as pd
from sklearn.datasets import load_iris
from google.colab import drive
import os

In [22]:
drive.mount('/content/drive')

Mounted at /content/drive


In [25]:
os.makedirs(
    "/content/drive/MyDrive/Projeto_ANN",
    exist_ok=True
)

In [21]:
iris = load_iris()

In [3]:
df = pd.DataFrame(
    iris.data,
    columns=iris.feature_names
)

df["target"] = iris.target

df.head()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0


## 2. Limpeza dos dados

Nesta etapa será realizada a análise e limpeza do dataset Iris, verificando
valores ausentes, registros duplicados, tipos de dados, valores inválidos
e a distribuição das classes.

### 2.1 Verificação do tamanho do dataset

Inicialmente, foi verificado o tamanho do dataset utilizando o atributo
`shape`, que informa a quantidade de linhas e colunas existentes na base.

In [4]:
df.shape

(150, 5)

### 2.2 Verificação dos tipos de dados

O método `info()` foi utilizado para verificar os tipos das variáveis e
identificar possíveis valores ausentes no dataset.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


### 2.3 Verificação de valores ausentes

In [6]:
df.isnull().sum()

,0
sepal length (cm),0
sepal width (cm),0
petal length (cm),0
petal width (cm),0
target,0


Foi realizada a verificação de valores nulos no dataset Iris. Não foram encontrados registros com dados ausentes, portanto não foi necessário realizar preenchimento ou remoção de linhas.

### 2.5 Verificação de registros duplicados

Registros duplicados podem causar redundância nos dados e influenciar
desnecessariamente o treinamento do modelo.

Foi utilizado o método `duplicated()` para identificar possíveis registros
repetidos.

In [9]:
duplicados = df[df.duplicated()]

duplicados

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
142,5.8,2.7,5.1,1.9,2


In [10]:
df.duplicated().sum()

np.int64(1)

In [11]:
df = df.drop_duplicates()

### 2.6 Verificação do tamanho após a remoção de duplicatas

Após a remoção do registro duplicado, foi verificado novamente o tamanho
do dataset.

In [13]:
df.duplicated().sum()

np.int64(0)

In [14]:
df.shape

(149, 5)

Durante a etapa de limpeza dos dados, foi identificado 1 registro duplicado e foi decidido removê-la para evitar redundância.
Para removê-lo, foi utilizado o método `drop_duplicates()`, reduzindo o dataset de 150 para 149 registros.

### 2.7 Análise estatística dos dados

O método `describe()` foi utilizado para analisar medidas estatísticas
como média, desvio padrão, valores mínimos, máximos e quartis.

Essa análise auxilia na identificação de possíveis valores inconsistentes
ou fora do intervalo esperado.

In [15]:
df.describe()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
count,149.000000,149.000000,149.000000,149.000000,149.000000
mean,5.843624,3.059732,3.748993,1.194631,0.993289
std,0.830851,0.436342,1.767791,0.762622,0.817847
min,4.300000,2.000000,1.000000,0.100000,0.000000
25%,5.100000,2.800000,1.600000,0.300000,0.000000
50%,5.800000,3.000000,4.300000,1.300000,1.000000
75%,6.400000,3.300000,5.100000,1.800000,2.000000
max,7.900000,4.400000,6.900000,2.500000,2.000000


A análise estatística não apresentou valores claramente inválidos ou
incompatíveis com as medidas das flores presentes no dataset Iris.

### 2.8 Verificação de valores inválidos

As quatro características do dataset representam medidas de comprimento
em centímetros. Por esse motivo, valores negativos seriam considerados
inválidos.

Foi realizada uma verificação para identificar a existência desses valores.

In [16]:
(df.iloc[:, :-1] < 0).sum()

,0
sepal length (cm),0
sepal width (cm),0
petal length (cm),0
petal width (cm),0


Não foram encontrados valores negativos nas características analisadas.
Portanto, não foi necessário remover ou corrigir registros nesta etapa.

### 2.9 Verificação da distribuição das classes

A coluna `target` representa a espécie de cada flor.

Os valores são:

- `0` = Setosa
- `1` = Versicolor
- `2` = Virginica

Foi verificada a quantidade de registros pertencentes a cada classe.

In [17]:
df["target"].value_counts()

,count
target,
0,50
1,50
2,49


O dataset apresenta uma distribuição bastante equilibrada entre as três
classes, mesmo após a remoção do registro duplicado.

### 2.10 Conferência final

Após a realização da limpeza, foi feita uma última verificação dos dados
para confirmar a ausência de valores nulos e registros duplicados.

In [18]:
print("Tamanho final:", df.shape)

print("\nValores ausentes:")
print(df.isnull().sum())

print("\nRegistros duplicados:")
print(df.duplicated().sum())

print("\nDistribuição das classes:")
print(df["target"].value_counts())

Tamanho final: (149, 5)

Valores ausentes:
sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     0
target               0
dtype: int64

Registros duplicados:
0

Distribuição das classes:
target
0    50
1    50
2    49
Name: count, dtype: int64


### 2.11 Conclusão da limpeza

O dataset Iris apresentou poucos problemas relacionados à qualidade dos
dados. Não foram encontrados valores ausentes ou valores negativos.

Foi identificado e removido 1 registro duplicado, reduzindo o dataset de
150 para 149 amostras.

In [26]:
df.to_csv(
    "/content/drive/MyDrive/Projeto_ANN/iris_limpo.csv",
    index=False
)

## 3. Normalização/Padronização dos dados

Nesta etapa os dados foram padronizados para preparar o dataset para o
treinamento da rede neural. A padronização é importante em modelos de
ANN pois evita que features com escalas diferentes influenciem de forma
desigual o processo de aprendizado.

### 3.1 Separação entre features e target

A coluna `target` foi separada das demais, já que representa a classe
a ser prevista e não deve ser normalizada.

In [ ]:
X = df.drop(columns=["target"])
y = df["target"]

X.head()

### 3.2 Aplicação do StandardScaler

Foi utilizado o `StandardScaler` da biblioteca `scikit-learn`, que
transforma cada feature para ter média 0 e desvio padrão 1. Essa técnica
é amplamente utilizada em redes neurais por facilitar a convergência
do algoritmo de treinamento.

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

df_normalizado = pd.DataFrame(X_scaled, columns=X.columns)
df_normalizado["target"] = y.values

df_normalizado.head()

### 3.3 Verificação dos dados normalizados

Foi utilizado o método `describe()` para confirmar que as features
passaram a apresentar média próxima de 0 e desvio padrão próximo de 1.

In [ ]:
df_normalizado.describe()

### 3.4 Conclusão da normalização

As quatro features numéricas do dataset Iris foram padronizadas com
sucesso utilizando o `StandardScaler`, deixando todas as variáveis na
mesma escala. A coluna `target` foi mantida sem alterações, já que
representa a classe de cada registro.

In [ ]:
df_normalizado.to_csv(
    "/content/drive/MyDrive/Projeto_ANN/iris_normalizado.csv",
    index=False
)